run-cgprna/run_cgprna/fusion_tools.py

```diff
@@ -0,0 +1,173 @@
+DEFUSE_FILTER = Template('defuse_filters.pl -s $sample_name -o $out_dir -i $out_dir/${sample_name}.defuse-fusion.normals.filtered.txt')
+
+NORMAL_FUSION_ARG_NAME='-normals'
+
+REF_RELATED_DEFAULTS = {
+    'species': 'unspecified_species',
+    'ref_build': 'unspecified_ref_build',
+    'gene_build': 'ensembl'
+}
+
+
+def validate_input_seq_files(file_names):
+    '''
+    Perl wrappers: tophat_fusion.pl, star_fusion.pl and defuse.pl all support multiple input files but are restricted into a single type, i.e either BAM or FastQ files. As we're converting BAMs to fastqs before passing them to the wappers, we can easily handle a mixture of BAM and FastQ files, but needs to validate the files before starting to convert BAM to FastQ, so we don't waste time and resources on splitting a BAM when some of the FastQ files are not following the Perl wrapper's file name conventions.
+    '''
+    pairs = {}
+    for a_file in file_names:
+        if a_file.endswith('.bam'):
+            continue
+        if not os.path.exists(a_file):
+            sys.exit('Error: can not find input file: %s' % a_file)
+        a_file = os.path.abspath(a_file)
+        match = re.match(r'(.*)_([12])\.f(?:ast)?q(?:\.gz)?$', os.path.basename(a_file))
+        if match.group(2):
+            # if same mate number been spotted before or already got both mates
+            if pairs.get(match.group(1), 0) == int(match.group(2)) or pairs.get(match.group(1), 0) == 3:
+                sys.exit('Error: Too many \'_%s\' mate files for prefix: %s. Possibly redundant file: %s' % (match.group(2), match.group(1), a_file))
+            pairs[match.group(1)] = pairs.get(match.group(1), 0) + int(match.group(2))
+        else:
+            sys.exit('Error: File name does not follow expected coventions: %s' % a_file)
+    for a_mate, value in pairs.items():
+        # if both _1 and _2 of a pair of fastq files are give, value should be 3
+        if value != 3:
+            if value == 1:
+                sys.exit('Error: Can not find second mate file of: %s' % a_mate)
+            if value == 2:
+                sys.exit('Error: Can not find first mate file of: %s' % a_mate)
+            sys.exit('Error: Internal code logic error')
+
+
+
+# NOTE: So far only used if Defuse is given a gzipped fastq file
+def gunzip(source_file, dest_file):
+    print('unzipping %s ...' % os.path.basename(source_file), flush=True)
+    with gzip.open(source_file, 'rb') as s_file, open(dest_file, 'wb') as d_file:
+        shutil.copyfileobj(s_file, d_file, 65536)
+    print('done.', flush=True)
+
+
+def run_fusion_wrapper(args, temp_dir_name, fusion_templates, no_gzip=False):
+    # args to dict to allow updates later
+    args_dict = copy.deepcopy(vars(args))
+    # only use temp_dir when needed to extract reference files
+    temp_dir = os.path.join(os.path.abspath(args.out_dir), temp_dir_name)
+    mkdir(temp_dir)
```

Comment:
It will fail if the directory already exists. Capturing the exception would allow to send a clearer message to the user.